analyses/doublet-detection/exploratory-notebooks/03_compare-benchmark-results.Rmd

```diff
@@ -0,0 +1,363 @@
+doublet_df_list <- dataset_names |>
+  purrr::map(
+    \(dataset) {
+      
+      scdbl_tsv <- file.path(result_dir, glue::glue("{dataset}_scdblfinder.tsv"))
+      scrub_tsv <- file.path(result_dir, glue::glue("{dataset}_scrublet.tsv"))
+      sce_file <- file.path(data_dir, dataset, glue::glue("{dataset}_sce.rds"))
+      
+      scdbl_df <- scdbl_tsv |>
+        readr::read_tsv(show_col_types = FALSE) |>
+        dplyr::select(
+          barcodes,
+          cxds_score, 
+          scdbl_score = score, 
+          scdbl_prediction  = class
+        ) |>
+        # add cxds calls at 0.75 threshold
+        dplyr::mutate(
+          cxds_prediction = dplyr::if_else(
+            cxds_score >= 0.75,
+            "doublet",
+            "singlet"
+          )
+        ) 
+      
+      scrub_df <- readr::read_tsv(scrub_tsv, show_col_types = FALSE) 
+
+      # grab ground truth and PCA coordinates
+      sce <- readr::read_rds(sce_file)
+      sce_df <- scuttle::makePerCellDF(sce, use.dimred = "PCA") |>
+        tibble::rownames_to_column(var = "barcodes") |>
+        dplyr::select(barcodes,
+                      ground_truth = ground_truth_doublets, 
+                      PC1 = PCA.1, 
+                      PC2 = PCA.2)
+      rm(sce)
+      
+      dataset_df <- scdbl_df |>
+        dplyr::left_join(
+          scrub_df, 
+          by = "barcodes"
+        ) |>
+        dplyr::left_join(
+          sce_df, 
+          by = "barcodes"
+        ) 
+      
+      # Add a consensus call column
+      dataset_df <- dataset_df |>
+        dplyr::rowwise() |>
+        dplyr::mutate(consensus_call = dplyr::if_else(
+          all(
+            c(scdbl_prediction, scrublet_prediction, cxds_prediction) == "doublet"
+          ),
+          "doublet", 
+          "singlet"
+        )) |>
+        dplyr::mutate(
+          call_type = dplyr::case_when(
+            consensus_call == "doublet" && ground_truth == "doublet" ~ "tp",
+            consensus_call == "singlet" && ground_truth == "singlet" ~ "tn",
+            consensus_call == "doublet" && ground_truth == "singlet" ~ "fp",
+            consensus_call == "singlet" && ground_truth == "doublet" ~ "fn"
+          ), 
+          # set associated plotting colors
+          call_type_color = dplyr::case_when(
+            call_type == "tp" ~ unname(confusion_colors["tp"]),
+            call_type == "tn" ~ unname(confusion_colors["tn"]),
+            call_type == "fp" ~ unname(confusion_colors["fp"]),
+            call_type == "fn" ~ unname(confusion_colors["fn"])
+          )
+          )
+      
+      return(dataset_df)
+    }
+  )
+names(doublet_df_list) <- dataset_names
+```
+
+
+
+
+## Upset plots
+
+This section contains upset plots that show overlap across doublet calls from each method, displayed for each dataset.
+
+```{r}
+pull_barcodes <- function(df, pred_var) {
+  # Helper function to pull out barcodes for doublet calls
+  df$barcodes[df[[pred_var]] == "doublet"]
+}
```

Comment:
I don't know that having a function here is totally necessary since it's just one line